### ISSUE TITLE
preprocessing_function for image_data_generator()

### ISSUE THREAD

**Issue by paulofelipe** [rstudio/keras#225]
I'm trying to run the code below:

```r
# Finetune VGG16 Cats vs Dogs

library(keras)
library(dplyr)

# Preprocessing function --------------------------------------------------

preprocess_input <- function(x){
  
  x <- x[, , 3:1] #RGB -> BGR
  x[, , 1] <- x[, , 1] - 103.939
  x[, , 2] <- x[, , 2] - 116.779
  x[, , 3] <- x[, , 3] - 123.68
  return(x)
}

# Load VGG 16 ----------------------------------------------------------------
base_model <- application_vgg16(weights = 'imagenet', include_top = FALSE, 
                                input_shape = c(150, 150, 3))
summary(base_model)

# Top model
input_dim <- base_model$output %>% dim %>% unlist 
top_model <- keras_model_sequential() %>% 
  layer_flatten(input_shape = input_dim) %>% 
  layer_dense(units = 512, activation = "relu") %>% 
  layer_dense(units = 2, activation = "softmax")

model <- keras_model(input = base_model$input,
                     outputs = top_model(base_model$output))
  
freeze_weights(model, to = "block5_pool")
summary(model)
# Images generator ----------------------------------------------------------
batch_size <- 32

img_gen <- image_data_generator(preprocessing_function = preprocess_input)

train_gen <- flow_images_from_directory(directory = 'dados/dogscats/train/',
                                        generator = img_gen,
                                        target_size = c(150, 150),
                                        batch_size = batch_size,
                                        class_mode =  "categorical")

val_gen <- flow_images_from_directory(directory = 'dados/dogscats/valid/',
                                      generator = img_gen,
                                      target_size = c(150, 150),
                                      batch_size = batch_size,
                                      class_mode =  "categorical")

model %>% compile(
  optimizer = "rmsprop",
  loss = "binary_crossentropy",
  metrics = "accuracy"
)

history <- model %>% fit_generator(
  train_gen, 
  steps_per_epoch = train_gen$n/batch_size,
  epochs = 10, 
  validation_data = val_gen
)
```

However, a I get an error message:
```
Erro: C stack usage  580180996948 is too close to the limit
```

If I don't set a `preprocessing_function`, the code runs  correctly.

This is my session info :
```
R version 3.4.3 (2017-11-30)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 17.04

Matrix products: default
BLAS: /usr/lib/libblas/libblas.so.3.7.0
LAPACK: /usr/lib/lapack/liblapack.so.3.7.0

locale:
 [1] LC_CTYPE=pt_BR.UTF-8       LC_NUMERIC=C               LC_TIME=pt_BR.UTF-8        LC_COLLATE=pt_BR.UTF-8     LC_MONETARY=pt_BR.UTF-8    LC_MESSAGES=pt_BR.UTF-8    LC_PAPER=pt_BR.UTF-8      
 [8] LC_NAME=C                  LC_ADDRESS=C               LC_TELEPHONE=C             LC_MEASUREMENT=pt_BR.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] dplyr_0.7.2      keras_2.1.2.9000

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.14          knitr_1.17            bindr_0.1             whisker_0.3-2         magrittr_1.5          devtools_1.12.0       R6_2.2.2              rlang_0.1.4          
 [9] httr_1.2.1            tools_3.4.3           git2r_0.18.0          withr_2.1.0.9000      tfruns_1.1            yaml_2.1.16           digest_0.6.12         assertthat_0.2.0     
[17] tibble_1.3.4          tensorflow_1.4.3.9000 bindrcpp_0.2          base64enc_0.1-3       curl_2.7              zeallot_0.0.6         memoise_1.1.0         glue_1.1.1           
[25] compiler_3.4.3        reticulate_1.3.1.9000 jsonlite_1.5          pkgconfig_2.0.1      ```

thanks in advance!

**Comment by jjallaire** [comment]
Thanks for reporting this! Fix is here: https://github.com/rstudio/keras/commit/c37ba69cd90d968445f63724e79d5a1ddce643d8

Note however that this fix will only work if you are running the latest version of the core Keras library (>= 2.1.2). You can check this with `keras:::keras_version()` and can update using `install_keras()`.

I am curious why you need the R preprocessing function at all here. Here's a blog post that describes transfer learning on cats and dogs (note we don't do any such preprocessing): https://tensorflow.rstudio.com/blog/keras-image-classification-on-small-datasets.html. This is in turn based on the following Python notebook which also doesn't do preprocessing of this sort: https://github.com/fchollet/deep-learning-with-python-notebooks/blob/master/5.2-using-convnets-with-small-datasets.ipynb.

**Comment by paulofelipe** [comment]
Thanks for the fixing! It worked!

So, I figured the question about the reason I am using this preprocessing function would be asked. I am using two main sources: Deep Learning with R and fast.ai course. In the fast.ai course, Jeremy Howard uses this preprocessing function. He argues that the VGG16 was trained using this preprocessing steps. I coded the example in python and this preprocessing function improved the model's accuracy. Using the full sample (23k for training and 2k for validation) it gets 0,97+ accuracy without data augmentation and only training one fully connected layer on top of the convolution (frozen) based model. 

I will try the your example (2000k samples) with data augmentation and including/not including the preprocessing to see what happens.

Lastly, thanks to make deep learning more accessible in R.

**Comment by jjallaire** [comment]
Okay, that's very interesting. Let me know what you find out! It would also be interesting to see how the code from the book / post fares if provided the entire sample.

**Comment by geraudster** [comment]
Hi, it seems that the issue is still there in Keras 2.2.0.
I've tried to launch the same code (after changing path to images), and got the error :
`Error: C stack usage  174396169332 is too close to the limit`

My session info:
```
R version 3.4.4 (2018-03-15)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 16.04.5 LTS

Matrix products: default
BLAS: /usr/lib/atlas-base/atlas/libblas.so.3.0
LAPACK: /usr/lib/atlas-base/atlas/liblapack.so.3.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=fr_FR.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=fr_FR.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=fr_FR.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=fr_FR.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] dplyr_0.7.6      keras_2.2.0.9001

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.19     lattice_0.20-35  crayon_1.3.4     assertthat_0.2.0
 [5] zeallot_0.1.0    grid_3.4.4       R6_2.3.0         jsonlite_1.5    
 [9] magrittr_1.5     pillar_1.3.0     tfruns_1.4       rlang_0.2.2     
[13] bindrcpp_0.2.2   whisker_0.3-2    Matrix_1.2-14    reticulate_1.10 
[17] generics_0.0.1   tools_3.4.4      glue_1.3.0       purrr_0.2.5     
[21] compiler_3.4.4   pkgconfig_2.0.2  base64enc_0.1-3  tensorflow_1.9  
[25] tidyselect_0.2.4 bindr_0.1.1      tibble_1.4.2
```
`keras:::keras_version()` gives 2.2.4

**Comment by skeydan** [comment]
Thanks for the information! Will be fixed in 

https://github.com/rstudio/keras/pull/595

Here, just for further reference, is some simplified testing code. 
Also, seeing the discussion above, perhaps you want to use `imagenet_preprocess_input`?
(After this has been fixed I mean).

```
library(keras)

# Preprocessing function --------------------------------------------------

preprocess_input <- function(x){
  
  x <- x[, , 3:1] #RGB -> BGR
  x[, , 1] <- x[, , 1] - 103.939
  x[, , 2] <- x[, , 2] - 116.779
  x[, , 3] <- x[, , 3] - 123.68
  x
}

# Load VGG 16 ----------------------------------------------------------------

base_model <- application_vgg16(weights = 'imagenet', include_top = FALSE, 
                                input_shape = c(150, 150, 3))

model <- keras_model_sequential() %>%
  base_model %>%
  layer_flatten() %>% 
  layer_dense(units = 512, activation = "relu") %>% 
  layer_dense(units = 2, activation = "softmax")

freeze_weights(base_model)


# Images generator ----------------------------------------------------------

batch_size <- 6

img_gen <- image_data_generator(preprocessing_function = imagenet_preprocess_input)
#img_gen <- image_data_generator(preprocessing_function = preprocess_input)
#img_gen <- image_data_generator(preprocessing_function = NULL)


train_gen <- flow_images_from_directory(directory = 'train/',
                                        generator = img_gen,
                                        target_size = c(150, 150),
                                        batch_size = batch_size,
                                        class_mode =  "categorical")

model %>% compile(
  optimizer = "rmsprop",
  loss = "categorical_crossentropy",
  metrics = "accuracy"
)

history <- model %>%fit_generator(
  train_gen, 
  #steps_per_epoch = train_gen$n/batch_size,
  steps_per_epoch = 10,
  epochs = 1
)

```

**Comment by geraudster** [comment]
For sure, it was my first idea, but my code sample was not as clean as the one in this issue...

Thanks for your reply !

**Comment by skeydan** [comment]
it's fixed now, if you install from github :-)

`devtools::install_github("rstudio/keras")`

**Comment by DSLituiev** [comment]
I am running in the same issue with later versions of keras and R.

 I did `devtools::install_github("rstudio/keras")`, and I am using the code by @skeydan (Nov 9, 2018) with slight modification for my own input. 

Getting: `Error: C stack usage  71473204052 is too close to the limit` and kernel crash

```
> sessionInfo()
R version 3.6.0 (2019-04-26)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 16.04.6 LTS

Matrix products: default
BLAS:   /usr/lib/libblas/libblas.so.3.6.0
LAPACK: /usr/lib/lapack/liblapack.so.3.6.0

locale:
 [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       
 [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   
 [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          
[10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] keras_2.2.4.1.9001

loaded via a namespace (and not attached):
 [1] compiler_3.6.0    Matrix_1.2-17     magrittr_1.5      R6_2.4.0         
 [5] generics_0.0.2    tools_3.6.0       whisker_0.3-2     base64enc_0.1-3  
 [9] Rcpp_1.0.1        reticulate_1.12   tensorflow_1.13.1 grid_3.6.0       
[13] zeallot_0.1.0     jsonlite_1.6      tfruns_1.4        lattice_0.20-38  
```

**Comment by jayqi** [comment]
In case anyone else has run into this recently, I had the same error while using python module `keras_preprocessing` 1.1.0.

Turns out the fix is on master [here](https://github.com/rstudio/keras/blob/36309b34f9875a8c9f4b84d7f442ce5e8a7c4660/R/model.R#L1030-L1031) already, but is not in R package keras 2.2.4.1 which is the latest available release on CRAN.

I was able to have it work correctly by installing from source from GitHub.

**Comment by Cdk29** [comment]
I came accross the same error today, on a [kaggle notebook here]( https://www.kaggle.com/cdk292/efficientnet-b7-optimising-tta-with-r-and-keras).

Running : 

```
preprocess_input <- function(tensor){

    tensor<-keras::k_reverse(x=tensor, axes=2)

  return(tensor)
}

datagen <- image_data_generator(
   preprocessing_function = preprocess_input,
)

tta_generator <- flow_images_from_dataframe(dataframe = data, 
                                              directory = image_path,
                                              generator = datagen,
                                              class_mode = NULL,
                                              x_col = "image_id",
                                              y_col = NULL,
                                              target_size = c(600, 600),
                                              shuffle = FALSE,
                                              batch_size=1)

```

Gave me :

> Error: C stack usage  645494910132 is too close to the limit

I am clueless about the origin of this issue.
